### PR TITLE
Cmake lexer spaces in args

### LINKED ISSUE #5435: cmake lexer/parser doesn't understand spaces
I've been running into problems trying to build mesa on windows with llvm using the cmake finder for llvm. It turns out that having spaces in path names breaks the assumption of the cmake parser/lexer that spaces are always between the arguments of a function. Which isn't true because the replace tool doesn't add quotes around string values, so we'll need to try to figure that out ourselves.

### PR DESCRIPTION
**Pull request by dcbaker**
There is a rather nasty/annoying bug in the cmake parser (described in the bug linked) that breaks using cmake on windows in basically every case. I've added a big mean test case and fixed at least enough to make this work for me (mesa + llvm on windows), however, there may be more cases that need fixing.

Fixes #5435

### CODE REVIEW

File: mesonbuild/dependencies/base.py [hunk 1]
```diff
@@ -1419,14 +1419,14 @@ def _detect_dep(self, name: str, modules: List[Tuple[str, bool]], args: List[str
...
+        if not modules:
             incDirs = self.get_first_cmake_var_of(['PACKAGE_INCLUDE_DIRS'])
             defs = self.get_first_cmake_var_of(['PACKAGE_DEFINITIONS'])
             libs = self.get_first_cmake_var_of(['PACKAGE_LIBRARIES'])
 
             # Try to use old style variables if no module is specified
-            if len(libs) > 0:
-                self.compile_args = list(map(lambda x: '-I{}'.format(x), incDirs)) + defs
+            if libs:
+                self.compile_args = ['-I{}'.format(d) for d in incDir] + defs
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ [F821] undefined name 'incDir' ([link](https://sider.review/gh/repos/19784232/pulls/5436#issue-4149517))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
I pulled the cleanup patches out, I'll re-send those separately

**Comment by dcbaker** [comment]
I don't understand why the rust tests on windows have randomly started failing?

**Comment by dcbaker** [comment]
I'll see if I can reproduce the rust thing.

@mensinda, any comments?

**Comment by mensinda** [comment]
I am not sure how this could affect rust, but I do see another potential problem here (had no time to test it yet though). The CMake set function can also initialize lists like this:

From the docs:
> Signatures of this command that specify a <value>... placeholder expect zero or more arguments. Multiple arguments will be joined as a semicolon-separated list to form the actual variable value to be set. Zero arguments will cause normal variables to be unset. See the unset() command to unset variables explicitly.

```cmake
set(LIST a b c d)
set(STR "a b c d")
```

Unfortunately, there is no way to distinguish these to cases from debug trace. I thought about parsing the actual CMake file to figure out what the parameters are, but this seems like a lot of work...

I am not sure what to do about this.

**Comment by mensinda** [comment]
Ah, nevermind. I found a solution for the `set` function problem: In the CMake trace CMake will insert `;` for lists if a variable is used:

```cmake
cmake_minimum_required(VERSION 3.5)

set(MY_LIST a b c "d 3")
set(D2 ${MY_LIST})
```

produced this trace:

```
/home/daniel/temp/test/cmake/CMakeLists.txt(1):  cmake_minimum_required(VERSION 3.5 )
/home/daniel/temp/test/cmake/CMakeLists.txt(3):  set(MY_LIST a b c d 3 )
/home/daniel/temp/test/cmake/CMakeLists.txt(4):  set(D2 a;b;c;d 3 )
```

So, we just have to do `set(X ${X})` in our custom `CMakeLists.txt` and we should be fine. I think we are already doing that anyway. This way, your changes should be fine. However, it would be really nice if CMake wouldn't swallow all the quotation marks :(

**Comment by dcbaker** [comment]
Yeah, it would be really nice if cmake didn't swallow quotes.

**Comment by dcbaker** [comment]
I think the rust thing may be a rust change.

**Comment by jpakkane** [comment]
There's also this:

```
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.7.3\x64\lib\concurrent\futures\process.py", line 232, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "D:\a\1\s\run_project_tests.py", line 321, in run_test
    mlog.shutdown() # Close the log file because otherwise Windows wets itself.
  File "D:\a\1\s\run_project_tests.py", line 82, in __exit__
    mesonlib.windows_proof_rmtree(self.dir)
  File "D:\a\1\s\mesonbuild\mesonlib.py", line 1254, in windows_proof_rmtree
    shutil.rmtree(f)
  File "C:\hostedtoolcache\windows\Python\3.7.3\x64\lib\shutil.py", line 513, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\hostedtoolcache\windows\Python\3.7.3\x64\lib\shutil.py", line 392, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\hostedtoolcache\windows\Python\3.7.3\x64\lib\shutil.py", line 397, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\hostedtoolcache\windows\Python\3.7.3\x64\lib\shutil.py", line 395, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\a\\1\\s\\b 3734637e31\\meson-logs\\testlog.json'
```

Which is weird. It should not be doing that. Maybe someone is holding on to a file they should not?

**Comment by mensinda** [comment]
@jpakkane This is most likely a CI issue (again). [This job](https://dev.azure.com/jussi0947/jussi/_build/results?buildId=2668) for #5393 is also failing with the same errors as far as I can tell. I guess its time to debug Azure...

**Comment by jpakkane** [comment]
Ignoring the CI failure, is this ready to merge?

**Comment by mensinda** [comment]
@jpakkane From my point of view, yes.

**Pull request merged**